main/lsp/LSPTypechecker.cc

```diff
@@ -390,141 +331,214 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
+                case SlowPathMode::Init: {
+                    ENFORCE(!this->initialized);
+                    timeit.emplace(this->config->logger, "initial_index");
 
-                {
-                    Timer timeit(config->logger, "reIndexFromFileSystem");
-
-                    auto inputFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
-                    this->indexed.clear();
-                    this->indexed.resize(finalGS->filesUsed());
-
-                    auto asts = hashing::Hashing::indexAndComputeFileHashes(*finalGS, config->opts, *config->logger,
-                                                                            absl::Span<core::FileRef>(inputFiles),
-                                                                            workers, ownedKvstore);
-                    ENFORCE(asts.hasResult(), "LSP initialization does not support cancellation");
-                    // asts are in fref order, but we (currently) don't index and compute file hashes for payload files,
-                    // so vector index != FileRef ID. Fix that by slotting them into `indexed`.
-                    for (auto &ast : asts.result()) {
-                        int id = ast.file.id();
-                        ENFORCE_NO_TIMER(id < this->indexed.size());
-                        this->indexed[id] = std::move(ast);
+                    this->workspaceFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
+                    break;
+                }
+
+                // Reindexing on the slow path derives the list of inputs files from the file table
+                case SlowPathMode::Cancelable: {
+                    timeit.emplace(this->config->logger, "slow_path_reindex");
+
+                    ENFORCE(updates.updatedFiles.size() == updates.updatedFileIndexes.size());
+
+                    // Move the indexed trees into our local cache for open files. We will still reindex them again to
+                    // force any errors, but this is a convenient way to pre-populate our cache.
+                    if (!updates.updatedFiles.empty()) {
+                        for (auto &ast : updates.updatedFileIndexes) {
+                            if (ast.tree) {
+                                updates.updatedFinalGSFileIndexes.emplace_back(std::move(ast));
+                            }
+                        }
+
+                        updates.updatedFiles.clear();
+                        updates.updatedFileIndexes.clear();
+                    }
+
+                    this->workspaceFiles.clear();
+                    this->workspaceFiles.reserve(finalGS->filesUsed());
+
+                    // Rebuild the set of filerefs we're going to index.
+                    auto ix = -1;
+                    for (const auto &file : finalGS->getFiles()) {
+                        ++ix;
+
+                        if (file == nullptr) {
+                            continue;
+                        }
```

Comment:
Which files are `nullptr` in the file table?

It seems like possibly it's just the first element in the `GlobalState::files` vector, which corresponds to `FileRef id=0`, the file that doesn't exist.

If that's case, and because we've already changed `getFiles()` to return an `absl::Span` instead of a `const vector<...> &`, it seems like it would be pretty easy to return a span that skips the first one, and then we would be able to assert that the `file` is never nullptr.

Reply:
That would be a really nice change! I'll make that as a separate PR.

Reply:
Ah, one problem is that we index into that result with `core::FileRef::id()` in many places. One thing we could do is locally skip the first element here by chaining a `subspan(1)` on the end of `getFiles`, and initializing `ix` to `0` instead of `-1`. I'll give that a try and see if anything breaks.